### PR TITLE
Add JeProfService to allow dumping of Jemalloc heap profiles during running.

### PR DESCRIPTION
**Pull request by gartung**
Add JeProfService to allow dumping of Jemalloc heap profiles during running.

### CODE REVIEW

**Comment by gartung** [comment]
hold

**Comment by makortel** [comment]
Patrick asked elsewhere about a placement of this code. Dependence-wise the `IgTools/IgProf` would be ok, but logically e.g.  `PerfTools/JeProf` would sounds better (and makes me think if we should move `IgProf` from `IgTools` to `PerfTools`). @smuzaffar what do you think?

**Comment by gartung** [comment]
@makortel I redid the initial commit and put the files where you suggested.

**Comment by smuzaffar** [comment]
I agree @makortel , `PerfTools/JeProf` seems better.

**Comment by gartung** [comment]
@makortel I am getting heap profile dumps after 1, int((N-1)/2) and N-1 events just like with IgProfService. I think
this is ready for testing.

**Comment by gartung** [comment]
The `jeprof` perl executable is not doing the cumulative sums correctly. It is based on `pprof` which also does not understand how to handle the TBB task swapping.

**Comment by gartung** [comment]
However the heap dumps should be usable by `igprof-analyze` once I figure out how to read the heap dump and output an igprof formatted file.

**Comment by makortel** [comment]
@cmsbuild, please test

File: PerfTools/JeProf/plugins/JeProfService.cc [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+extern "C" {
+typedef int (*mallctl_t)(const char *name, void *oldp, size_t *oldlenp, void *newp, size_t newlen);
+}
+
+namespace {
+  bool initialize_prof();
+
+  mallctl_t mallctl = nullptr;
+  const bool have_jemalloc_and_prof = initialize_prof();
```
Comment (makortel) [inline PerfTools/JeProf/plugins/JeProfService.cc:29]: Could these become member variables of the `JeProfService`?

File: PerfTools/JeProf/plugins/JeProfService.h [hunk 2]
```diff
@@ -0,0 +1,91 @@
+#ifndef IgTools_JeProf_JeProfService_h
+#define IgTools_JeProf_JeProfService_h
```
Comment (makortel) [inline PerfTools/JeProf/plugins/JeProfService.h:2]: Could you move the contents of this header to the `.cc` file?

File: PerfTools/JeProf/plugins/JeProfService.cc (hunk 1 above)
Comment (gartung) [reply PerfTools/JeProf/plugins/JeProfService.cc:29]: Just copying the pattern the was used here
https://github.com/cms-sw/cmssw/blob/master/HLTrigger/Timer/src/memory_usage.cc

File: PerfTools/JeProf/plugins/JeProfService.h (hunk 2 above)
Comment (gartung) [reply PerfTools/JeProf/plugins/JeProfService.h:2]: OK. I was just following the pattern or IgProfService.

**Comment by gartung** [comment]
please test

**Comment by gartung** [comment]
@makortel This is ready to be merged. Can you please unhold it.

**Comment by makortel** [comment]
unhold

**Comment by makortel** [comment]
The package is added for bot in https://github.com/cms-sw/cms-bot/pull/1929

**Comment by makortel** [comment]
ping bot

**Comment by makortel** [comment]
+1

**Comment by perrotta** [comment]
+1

**Pull request merged**